### PR TITLE
OSHMEM v1.3: c11 generics (DNM before #2103)

### PR DESCRIPTION
**Pull request by alex-mikheev**
@jladd-mlnx @igor-ivanov please review
Note that PR depends on https://github.com/open-mpi/ompi/pull/2103

### CODE REVIEW

File: oshmem/include/pshmem.h [hunk 1]
```diff
@@ -111,13 +143,29 @@ OSHMEM_DECLSPEC  void pshmem_putmem(void *target, const void *source, size_t len
...
+#define pshmem_iput(dst, src, tst, sst, len, pe)     \
+    _Generic(&*(dst),                                \
+            char*:        pshmem_char_iput,          \
+            short*:       pshmem_short_iput,         \
+            int*:         pshmem_int_iput,           \
+            long*:        pshmem_long_iput,          \
+            long long*:   pshmem_longlong_iput,      \
+            float*:       pshmem_float_iput,         \
+            double*:      pshmem_double_iput,        \
+            long double*: pshmem_longdouble_iput)(dst, src, len, pe)
```
Comment (bertwesarg) [inline oshmem/include/pshmem.h]: This call does not match the macros arguments.

File: oshmem/include/pshmem.h [hunk 2]
```diff
@@ -151,17 +212,45 @@ OSHMEM_DECLSPEC  float pshmem_float_g(const float* addr, int pe);
...
+#define pshmem_g(addr, pe)                                    \
+    _Generic(&*(dst),                                         \
+            char*:        pshmem_char_g,                      \
+            short*:       pshmem_short_g,                     \
+            int*:         pshmem_int_g,                       \
+            long*:        pshmem_long_g,                      \
+            long long*:   pshmem_longlong_g,                  \
+            float*:       pshmem_float_g,                     \
+            double*:      pshmem_double_g,                    \
+            long double*: pshmem_longdouble_g)(addr, val, pe)
```
Comment (bertwesarg) [inline oshmem/include/pshmem.h]: Dito.

File: oshmem/include/shmem.h.in [hunk 3]
```diff
@@ -157,40 +160,84 @@ OSHMEM_DECLSPEC  void shmem_float_p(float* addr, float value, int pe);
...
+#define shmem_iput(dst, src, tst, sst, len, pe)     \
+    _Generic(&*(dst),                               \
+            char*:        shmem_char_iput,          \
+            short*:       shmem_short_iput,         \
+            int*:         shmem_int_iput,           \
+            long*:        shmem_long_iput,          \
+            long long*:   shmem_longlong_iput,      \
+            float*:       shmem_float_iput,         \
+            double*:      shmem_double_iput,        \
+            long double*: shmem_longdouble_iput)(dst, src, len, pe)
```
Comment (bertwesarg) [inline oshmem/include/shmem.h.in]: Dito.

File: oshmem/include/shmem.h.in [hunk 4]
```diff
@@ -216,17 +277,45 @@ OSHMEM_DECLSPEC  float shmem_float_g(const float* addr, int pe);
...
+#define shmem_g(addr, pe)                                    \
+    _Generic(&*(dst),                                        \
+            char*:        shmem_char_g,                      \
+            short*:       shmem_short_g,                     \
+            int*:         shmem_int_g,                       \
+            long*:        shmem_long_g,                      \
+            long long*:   shmem_longlong_g,                  \
+            float*:       shmem_float_g,                     \
+            double*:      shmem_double_g,                    \
+            long double*: shmem_longdouble_g)(addr, val, pe)
```
Comment (bertwesarg) [inline oshmem/include/shmem.h.in]: Dito.

File: oshmem/include/pshmem.h [hunk 5]
```diff
@@ -92,17 +96,45 @@ OSHMEM_DECLSPEC  void pshmem_float_p(float* addr, float value, int pe);
...
+#define pshmem_p(dst, val, pe)                                \
+    _Generic(&*(dst),                                         \
+            char*:        pshmem_char_p,                      \
+            short*:       pshmem_short_p,                     \
+            int*:         pshmem_int_p,                       \
+            long*:        pshmem_long_p,                      \
+            long long*:   pshmem_longlong_p,                  \
+            float*:       pshmem_float_p,                     \
+            double*:      pshmem_double_p,                    \
+            long double*: shmem_longdouble_p)(dst, val, pe)
```
Comment (bertwesarg) [inline oshmem/include/pshmem.h]: that should be `pshmem_longdouble_p`

**Comment by jsquyres** [comment]
@alex-mikheev Thanks for the fixes.  Can you squash down those 2 "fix" comments into their appropriate original commit(s)?  (i.e., no need to have a PR that adds fixes to commits that were introduced in the same PR)

**Comment by alex-mikheev** [comment]
sure i will squash 'fixes commits'

**Comment by jladd-mlnx** [comment]
Merging this.

**Pull request merged**